File: custom_stock/custom_stock/report/stock_inventory/stock_inventory.js
// Copyright (c) 2016, BM and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Stock Inventory"] = {
	"filters": [
		{
			"fieldname": "warehouse",
			"label": __("Warehouse"),
			"fieldtype": "Link",
			"options": "Warehouse",
		},
		{
			"fieldname": "inventory_voucher",
			"label": __("Inventory Voucher"),
			"fieldtype": "Link",
			"options": "Stock Inventory Processing",
			"reqd": 1,
			"get_query": function () {
				let warehouse = frappe.query_report.get_filter_value("warehouse");
				let filters = {
					"docstatus": 1
				}
				if (warehouse) {
					filters.warehouse = warehouse;
				}
				return { "filters": filters }
			}
		},
		{
			"fieldname": "item_group",
			"label": __("Item Group"),
			"fieldtype": "MultiSelectList",
			get_data: function (txt) {
				if (!frappe.query_report.filters) return;
				let inventory_voucher = frappe.query_report.get_filter_value("inventory_voucher");
				// let filters = frappe.db.get_doc("Stock Inventory Processing", inventory_voucher).then((r) => {
				// 	console.log(r);
				// 	// if (r.item_group != null) {
				// 	// return frappe.db.get_link_options("Item Group", txt, { name: r.item_group });
				// 	// } else {
				// 	// return frappe.db.get_link_options("Item Group", txt);
				// 	// }
				// });
				// console.log(filters);
				if (!inventory_voucher) return;
				return frappe.db.get_link_options("Item Group", txt);
			}
		},
		{
			"fieldname": "not_in_item_group",
			"label": __("Not in Item Group"),
			"fieldtype": "MultiSelectList",
			get_data: function (txt) {
				if (!frappe.query_report.filters) return;
				let inventory_voucher = frappe.query_report.get_filter_value("inventory_voucher");
				if (!inventory_voucher) return;
				return frappe.db.get_link_options("Item Group", txt);
			}
		},
		{
			"fieldname": "inventory_type",
			"label": __("Inventory Type"),
			"fieldtype": "MultiSelectList",
			get_data: function (txt) {
				if (!frappe.query_report.filters) return;
				let inventory_voucher = frappe.query_report.get_filter_value("inventory_voucher");
				if (!inventory_voucher) return;
				let types = [
					{ value: __("Deficit"), description: __("Loss in stock") },
					{ value: __("Surplus"), description: __("Gain in stock") },
					{ value: __("Uninventorized"), description: __("Uninventorized Items") }
				]
				return types;
			}
		}
	]
};

File: custom_stock/custom_stock/report/stock_inventory/stock_inventory.py
# Copyright (c) 2013, BM and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _


def execute(filters=None):
    columns, data = get_columns(filters), get_data(filters)
    return columns, data


def get_data(filters):
    conditions, filters = get_conditions(filters)
    query = '''
		SELECT item_code, item_name, item_group, inventory_qty, actual_qty, difference_qty, rate, amount
		FROM `tabStock Inventory Processing Item`
		WHERE docstatus = 1 %s
	'''
    siis = frappe.db.sql(query % conditions, filters, as_dict=1)
    data = []
    for sii in siis:
        row = [sii.item_code, sii.item_name, sii.item_group,
               sii.inventory_qty, sii.actual_qty, sii.difference_qty, sii.rate, sii.amount]

        data.append(row)
    return data


def get_conditions(filters):
    conditions = ""
    # print(type(filters.get("inventory_type")))
    if filters.get("inventory_type"):
        # if len(filters.get("inventory_type")) > 1:
        #     connector = " or "
        # else:
        # connector = " and ("
        conditions += " and ("  # connector
        l = len(filters.get("inventory_type"))
        for it in filters.get("inventory_type"):
            l = l - 1
            if it == _("Deficit"):
                conditions += "(difference_qty < 0  and inventory_qty != 0)"
            if it == _("Surplus"):
                conditions += "difference_qty > 0"
            if it == _("Uninventorized"):
                conditions += "inventory_qty = 0"
            if l != 0:
                conditions += " or "
            else:
                conditions += ")"

    if filters.get("inventory_voucher"):
        conditions += " and parent = %(inventory_voucher)s"
    if filters.get("item_group"):
        conditions += " and item_group in %(item_group)s"
    if filters.get("not_in_item_group"):
        conditions += " and item_group not in %(not_in_item_group)s"

    return conditions, filters


def get_columns(filters):
    columns = [
        _("Item") + ":Link/Item:250",
        {
            "label": _("Item Name"),
            "fieldname": "item_name",
            "width": 200,
            "hidden": 1
        },
        _("Item Group") + ":Link/Item Group:200",
        _("Inventory Quantity") + "::100",
        _("Actual Quantity") + "::100",
        _("Difference Quantity") + "::100",
        _("Rate") + ":Currency:100",
        _("Amount") + ":Currency:100"
    ]

    return columns
